### ISSUE TITLE
What do you think about removing the controller hierarchy?

### ISSUE THREAD

**Issue by jacwright** [chip-js/chip#25]
One of the most confusing parts of working with Chip is the controller hierarchy. The benefit it provides is that you can add helpers and data at the top of your app and have them available everywhere.

The downside to controller hierarchy is that it is hard to reason about, to know what data is available, to know what controller has control at a given element, and to understand how the prototype chain works where you can "get" but not "set" etc.

The benefits I believe can be provided in other simpler ways. Filters can be used more widely as helpers as well, code access like `"filters.fullName(user) + ' Rocks!'"` if the `"user | fullName"` doesn't work for your use-case. Data can be passed in using the local binding, or retrieved some other way through injection (brainstorm here).

**Comment by matthopson** [comment]
In our case, wouldn't we lose some benefits of it as well though? For instance in a partial or bind-each where you expect the data to be scoped.

**Comment by jacwright** [comment]
Benefits would definitely be lost. The question is, can we meet those needs in other better ways?

A `bind-each` is the only place I expect to use the prototypal hierarchy, but it should be transparent to the developer in that case. The idea is that within a single template the same data/methods are available everywhere in it. A new template/controller is a new standalone (non-inheriting) scope. Data can be passed in as with the following example:

```html
<div bind-each="{{member in team.members}}>
  <div bind-partial="member-view" local-member="{{member}}"></div>
</div>
```

Speaking of, should we support light-weight custom elements?

```html
<div bind-each="{{member in team.members}}>
  <member-view member="{{member}}"></member-view>
</div>
```

**Comment by dmvt** [comment]
While I agree that the nested controllers can be hard to grok for beginners, I think it's a common enough approach these days that the learning curve is reasonable. In contrast, I think filters as cross controller objects could be confusing, since most frameworks use them as reducers. What you're describing seems to more closely match the Angular factory/service concept.

One of the things that I really like about Chip is that it marries the core binding and hierarchical concepts of Angular with the flexibility of backbone. I wouldn't want to see this disappear.

I do like the bind-partial simplification you describe. Seems like a simplification of the directive idea in Angular and a very solid potential core feature or addon.

**Comment by jacwright** [comment]
Done in https://github.com/chip-js/fragments-js, available in chip 2.0. We have found using mixins and a global data store that components are much easier to use and make a more modular application that is easier to follow.
